module/zfs/zio.c

```diff
@@ -143,12 +143,21 @@ zio_init(void)
 		 */
 		if (arc_watch && !IS_P2ALIGNED(size, PAGESIZE))
 			continue;
-#endif
+		/*
+		 * Here's the problem - on 4K native devices in userland on
+		 * Linux using O_DIRECT, buffers must be 4K aligned or I/O
+		 * will fail with EINVAL, causing zdb (and others) to coredump.
+		 * Since userland probably doesn't need optimized buffer caches,
+		 * we just force 4K alignment on everything.
+		 */
+		align = PAGESIZE;
```

Comment:
Maybe this should be 4096 explicitly, instead of `PAGESIZE`. 
XFS requires logical sector sector size of the target for alignment [link](http://lxr.free-electrons.com/source/fs/xfs/xfs_ioctl.c#L1671)

But how this translates to kernel slabs, is alignment required there as well?
Currently, `zio_buf` slabs are setup like this:
```
size: 512	align: 512	
size: 1024	align: 512	
size: 1536	align: 512	
size: 2048	align: 512	
size: 2560	align: 512	
size: 3072	align: 512	
size: 3584	align: 512	
size: 4096	align: 1024  <==
size: 5120	align: 1024
size: 6144	align: 1024
size: 7168	align: 1024
size: 8192	align: 2048
size: 10240	align: 2048
size: 12288	align: 2048
size: 14336	align: 2048
size: 16384	align: 4096
size: 20480	align: 4096
....		align: 4096
```